bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/handler/GardenaAccountHandler.java

```diff
@@ -79,52 +90,93 @@ public void setDiscoveryService(GardenaDeviceDiscoveryService discoveryService)
-                        webSocketFactory);
-                final GardenaDeviceDiscoveryService discoveryService = this.discoveryService;
-                if (discoveryService != null) {
-                    discoveryService.startScan(null);
-                    discoveryService.waitForScanFinishing();
-                }
-                updateStatus(ThingStatus.ONLINE);
-            } catch (GardenaException ex) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, ex.getMessage());
-                disposeGardena();
-                if (ex.getStatus() == 429) {
-                    // if there was an error 429 (Too Many Requests), wait for 24 hours before trying again
-                    scheduleReinitialize(REINITIALIZE_DELAY_HOURS_LIMIT_EXCEEDED, TimeUnit.HOURS);
-                } else {
-                    // otherwise reinitialize after 120 seconds
-                    scheduleReinitialize(REINITIALIZE_DELAY_SECONDS, TimeUnit.SECONDS);
+    private synchronized void initializeGardena() {
+        try {
+            GardenaConfig gardenaConfig = getThing().getConfiguration().as(GardenaConfig.class);
+            logger.debug("{}", gardenaConfig);
+
+            String id = getThing().getUID().getId();
+            gardenaSmart = new GardenaSmartImpl(id, gardenaConfig, this, scheduler, httpClientFactory,
+                    webSocketFactory);
+            final GardenaDeviceDiscoveryService discoveryService = this.discoveryService;
+            if (discoveryService != null) {
+                discoveryService.startScan(null);
+                discoveryService.waitForScanFinishing();
+            }
+            reInitializationCausedBy429 = false;
+            updateStatus(ThingStatus.ONLINE);
+        } catch (GardenaException ex) {
+            logger.warn("{}", ex.getMessage());
+
+            long secondsUntilRestart;
+
+            synchronized (reInitializationCodeLock) {
+                boolean isHttp429Error = (ex.getStatus() == 429);
+                boolean isRepeatError = lastErrorTime.plus(REINITIALIZE_DELAY_MINUTES_BACK_OFF)
+                        .isAfter(LocalDateTime.now());
+
+                Duration delay = isHttp429Error ? REINITIALIZE_DELAY_HOURS_LIMIT_EXCEEDED
+                        : isRepeatError ? REINITIALIZE_DELAY_MINUTES_BACK_OFF : REINITIALIZE_DELAY_SECONDS;
+
+                ScheduledFuture<?> reInitializationTask = this.reInitializationTask;
+                if (reInitializationTask == null || reInitializationTask.isDone()
+                        || (isHttp429Error != reInitializationCausedBy429)) {
+                    reInitializationTask = scheduleReinitialize(delay);
                 }
-                logger.warn("{}", ex.getMessage());
+                reInitializationCausedBy429 = isHttp429Error;
+                lastErrorTime = LocalDateTime.now();
```

Comment:
`LocalDateTime` is not well suited for this purpose as it's susceptible to clock changes. I would propose to use `Instant` instead.

Reply:
Ok. Done.